### PR TITLE
Idempotent start

### LINKED ISSUE #351: Stop failed to close currently tracking interval
Just taking a note so I do not forget to follow up later

With the current dev branch (cea2e5d13f26227063) I ran into a situation where I was unable to stop the project I was currently tracking. Below is reproducer:

```
$ rm -fr ~/.timewarrior/data
$ src/timew track :yesterday proja
Note: 'proja' is a new tag.
Recorded proja
  Started 2020-07-27T00:00:00
  Ended           28T00:00:00
  Total              24:00:00
$ src/timew start 08:00:00 projb
Note: 'projb' is a new tag.
Tracking projb
  Started 2020-07-28T08:00:00
  Current            11:15:38
  Total               3:15:38
$ src/timew annotate "Hello"
Annotated @1 with "Hello"
$ src/timew start projb
Tracking projb
  Started 2020-07-28T11:15:57
  Current                  57
  Total               3:15:57
$ src/timew export
[
{"id":3,"start":"20200727T050000Z","end":"20200728T050000Z","tags":["proja"]},
{"id":2,"start":"20200728T130000Z","tags":["projb"],"annotation":"Hello"},
{"id":1,"start":"20200728T161557Z","tags":["projb"]}
]
$ src/timew
Tracking projb
  Started 2020-07-28T11:15:57
  Current               20:17
  Total               3:24:37
$ src/timew stop
You cannot overlap intervals. Correct the start/end time, or specify the :adjust hint.
```

### PR DESCRIPTION
**Pull request by sruffell**
Closes #351 by restoring the behavior to 1.3.0.

That being said, I wonder if it would make more sense to produce an error if a user tries to run start multiple times? Currently, if someone runs start multiple times, the undo will not undo the start that silently failed, which seems a little surprising to me.

Maybe a start with the same tags should say "Already tracking xxxxx"  and then the :adjust hint could then mean stop the current interval and start a new one even if we're already tracking the same tags?  

However, since I'm not sure if taskwarrior depends on the previous behavior, I submit this change set which restores it.

### CODE REVIEW

**Comment by lauft** [comment]
I am sorry, but I have to reject this PR.

Yes, the behaviour of idempotent start has to be restored, but your PR fixes the problem at the database level which is too deep in my opinion.

In my analysis, the problem is rooted in the `autoAdjust` function line 111 where the function simply returns because the to be added interval is already tracked. `CmdStart` (and `CmdContinue`) ignores this and adds the interval anyway.

A simpler fix is to let `autoAdjust` and `validate` return a boolean which signals whether or not to add the interval. `CmdStart` and `CmdContinue` are the only commands, where this situation can occur, `CmdTrack` does not add open intervals and `CmdStop` does add open intervals but by design never with the same tag set as the currently open one.

This is hacky but I think contains the fix more than replacing the interval and handing the problem solution over to the database. In the long term the solution is to integrate `database.addInterval(...)` into `validate(...)` as these two always get called in pairs. However, this requires a some disentangling of the database and the access to it in beforehand.

File: src/validate.cpp [hunk 1]
```diff
@@ -104,11 +104,18 @@ static void autoAdjust (
   Interval latest = getLatestInterval (database);
   if (interval.is_open () && latest.encloses (interval))
   {
-    if (latest.tags () == interval.tags ())
+    if ((latest.tags () == interval.tags ()) &&
+        (interval.getAnnotation ().empty () ||
```
[the remark this answers is not in the document]
Comment (sruffell) [reply src/validate.cpp:108]: That is the way it is now, but there isn't anything that would preclude that in the future. i.e. #327

**Comment by lauft** [comment]
Heres the patch I plan to apply for the fix instead:
```
Index: test/start.t
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- test/start.t	(revision ea33b29986850d9007f2404d7e7030409db85fd4)
+++ test/start.t	(date 1597309886840)
@@ -131,11 +131,17 @@
 
     def test_start_with_same_tags_as_current_tracking(self):
         """Test 'start' with same tags as current tracking should not start new tracking"""
-        self.t("start 1h ago bar foo")
+        utc_now = datetime.now().utcnow()
+        one_hour_ago_utc = utc_now - timedelta(hours=1)
+
+        self.t("start {:%Y-%m-%dT%H:%M:%S}Z bar foo".format(one_hour_ago_utc))
 
-        code, out, err = self.t("start foo bar")
+        self.t("start foo bar")
 
-        self.assertNotIn("Recorded bar foo", out)
+        j = self.t.export()
+
+        self.assertEqual(len(j), 1)
+        self.assertOpenInterval(j[0], expectedStart=one_hour_ago_utc, expectedTags=["foo", "bar"])
 
     def test_single_interval_enclosing_exclusion(self):
         """Add one interval that encloses an exclusion, and is therefore flattened"""
Index: src/timew.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/timew.h	(revision ea33b29986850d9007f2404d7e7030409db85fd4)
+++ src/timew.h	(date 1597309652842)
@@ -59,7 +59,7 @@
 
 // validate.cpp
 void autoFill (const Rules&, Database&, Interval&);
-void validate (const CLI& cli, const Rules& rules, Database&, Interval&);
+bool validate (const CLI& cli, const Rules& rules, Database&, Interval&);
 
 // init.cpp
 bool lightweightVersionCheck (int, const char**);
Index: src/validate.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/validate.cpp	(revision ea33b29986850d9007f2404d7e7030409db85fd4)
+++ src/validate.cpp	(date 1597312345425)
@@ -91,7 +91,7 @@
 //   can involve rejection, adjustment of modified interval, or adjustment of
 //   recorded data.
 //
-static void autoAdjust (
+static bool autoAdjust (
   bool adjust,
   const Rules& rules,
   Database& database,
@@ -108,7 +108,7 @@
     {
       // If the new interval tags match those of the currently open interval,
       // then do nothing - the tags are already being tracked.
-      return;
+      return false;
     }
 
     database.deleteInterval (latest);
@@ -128,7 +128,7 @@
 
   if (overlaps.empty ())
   {
-    return;
+    return true;
   }
 
   debug ("Input         " + interval.dump ());
@@ -195,10 +195,11 @@
       }
     }
   }
+  return true;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-void validate (
+bool validate (
   const CLI& cli,
   const Rules& rules,
   Database& database,
@@ -210,7 +211,7 @@
     autoFill (rules, database, interval);
   }
 
-  autoAdjust (findHint (cli, ":adjust"), rules, database, interval);
+  return autoAdjust (findHint (cli, ":adjust"), rules, database, interval);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
Index: test/continue.t
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- test/continue.t	(revision ea33b29986850d9007f2404d7e7030409db85fd4)
+++ test/continue.t	(date 1597312118781)
@@ -394,6 +394,70 @@
                                 expectedTags=["FOO"],
                                 description="continued interval")
 
+    def test_continue_with_id_is_idempotent(self):
+        """Verify that continuing with id is idempotent"""
+        now_utc = datetime.now().utcnow()
+
+        two_hours_before_utc = now_utc - timedelta(hours=2)
+        three_hours_before_utc = now_utc - timedelta(hours=3)
+        four_hours_before_utc = now_utc - timedelta(hours=4)
+        five_hours_before_utc = now_utc - timedelta(hours=5)
+
+        self.t("track FOO {:%Y-%m-%dT%H}:00:00Z - {:%Y-%m-%dT%H}:00:00Z".format(five_hours_before_utc, four_hours_before_utc))
+        self.t("track BAR {:%Y-%m-%dT%H}:00:00Z - {:%Y-%m-%dT%H}:00:00Z".format(four_hours_before_utc, three_hours_before_utc))
+        self.t("start FOO {:%Y-%m-%dT%H}:00:00Z".format(three_hours_before_utc))
+        self.t("continue @3 {:%Y-%m-%dT%H}:00:00Z".format(now_utc))
+
+        j = self.t.export()
+
+        self.assertEqual(len(j), 3)
+        self.assertClosedInterval(j[0],
+                                  expectedStart="{:%Y%m%dT%H}0000Z".format(five_hours_before_utc),
+                                  expectedEnd="{:%Y%m%dT%H}0000Z".format(four_hours_before_utc),
+                                  expectedTags=["FOO"],
+                                  description="third interval")
+        self.assertClosedInterval(j[1],
+                                  expectedStart="{:%Y%m%dT%H}0000Z".format(four_hours_before_utc),
+                                  expectedEnd="{:%Y%m%dT%H}0000Z".format(three_hours_before_utc),
+                                  expectedTags=["BAR"],
+                                  description="second interval")
+        self.assertOpenInterval(j[2],
+                                expectedStart="{:%Y%m%dT%H}0000Z".format(three_hours_before_utc),
+                                expectedTags=["FOO"],
+                                description="first interval")
+
+    def test_continue_with_tag_is_idempotent(self):
+        """Verify that continuing with id is idempotent"""
+        now_utc = datetime.now().utcnow()
+
+        two_hours_before_utc = now_utc - timedelta(hours=2)
+        three_hours_before_utc = now_utc - timedelta(hours=3)
+        four_hours_before_utc = now_utc - timedelta(hours=4)
+        five_hours_before_utc = now_utc - timedelta(hours=5)
+
+        self.t("track FOO {:%Y-%m-%dT%H}:00:00Z - {:%Y-%m-%dT%H}:00:00Z".format(five_hours_before_utc, four_hours_before_utc))
+        self.t("track BAR {:%Y-%m-%dT%H}:00:00Z - {:%Y-%m-%dT%H}:00:00Z".format(four_hours_before_utc, three_hours_before_utc))
+        self.t("start FOO {:%Y-%m-%dT%H}:00:00Z".format(three_hours_before_utc))
+        self.t("continue FOO {:%Y-%m-%dT%H}:00:00Z".format(now_utc))
+
+        j = self.t.export()
+
+        self.assertEqual(len(j), 3)
+        self.assertClosedInterval(j[0],
+                                  expectedStart="{:%Y%m%dT%H}0000Z".format(five_hours_before_utc),
+                                  expectedEnd="{:%Y%m%dT%H}0000Z".format(four_hours_before_utc),
+                                  expectedTags=["FOO"],
+                                  description="third interval")
+        self.assertClosedInterval(j[1],
+                                  expectedStart="{:%Y%m%dT%H}0000Z".format(four_hours_before_utc),
+                                  expectedEnd="{:%Y%m%dT%H}0000Z".format(three_hours_before_utc),
+                                  expectedTags=["BAR"],
+                                  description="second interval")
+        self.assertOpenInterval(j[2],
+                                expectedStart="{:%Y%m%dT%H}0000Z".format(three_hours_before_utc),
+                                expectedTags=["FOO"],
+                                description="first interval")
+
     def test_continue_with_id_and_range(self):
         """Verify that continue with a range adds a copy with same tags"""
         now_utc = datetime.now().utcnow()
Index: src/commands/CmdStart.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/commands/CmdStart.cpp	(revision ea33b29986850d9007f2404d7e7030409db85fd4)
+++ src/commands/CmdStart.cpp	(date 1597309652832)
@@ -51,10 +51,11 @@
   }
 
   journal.startTransaction ();
-  validate (cli, rules, database, interval);
-  database.addInterval (interval, verbose);
-  journal.endTransaction ();
-
+  if (validate (cli, rules, database, interval))
+  {
+    database.addInterval (interval, verbose);
+    journal.endTransaction ();
+  }
   if (verbose)
   {
     std::cout << intervalSummarize (database, rules, interval);
Index: src/commands/CmdContinue.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/commands/CmdContinue.cpp	(revision ea33b29986850d9007f2404d7e7030409db85fd4)
+++ src/commands/CmdContinue.cpp	(date 1597312118785)
@@ -114,10 +114,11 @@
   to_copy.end = end_time;
 
   journal.startTransaction ();
-  validate (cli, rules, database, to_copy);
-  database.addInterval (to_copy, verbose);
-  journal.endTransaction ();
-
+  if (validate (cli, rules, database, to_copy))
+  {
+    database.addInterval (to_copy, verbose);
+    journal.endTransaction ();
+  }
   if (verbose)
   {
     std::cout << intervalSummarize (database, rules, to_copy);
```

File: test/start.t [hunk 2]
```diff
@@ -271,6 +271,19 @@ class TestStart(TestCase):
         self.t("start 1h ago proja")
         code, out, err = self.t("start 2h ago proja :adjust")
 
+    def test_start_with_same_tags(self):
```
[the remark this answers is not in the document]
Comment (lauft) [reply test/start.t:274]: Also see the attached patch where I updated and added the tests for idempotent `start`/`continue`

**Comment by sruffell** [comment]
> I am sorry, but I have to reject this PR.

No problem.

> Yes, the behaviour of idempotent start has to be restored, but your PR fixes the problem at the database level which is too deep in my opinion.
>
> In my analysis, the problem is rooted in the `autoAdjust` function line 111 where the function simply returns because the to be added interval is already tracked. `CmdStart` (and `CmdContinue`) ignores this and adds the interval anyway.
> 
> A simpler fix is to let `autoAdjust` and `validate` return a boolean which signals whether or not to add the interval. `CmdStart` and `CmdContinue` are the only commands, where this situation can occur, `CmdTrack` does not add open intervals and `CmdStop` does add open intervals but by design never with the same tag set as the currently open one.
> 
> This is hacky but I think contains the fix more than replacing the interval and handing the problem solution over to the database. In the long term the solution is to integrate `database.addInterval(...)` into `validate(...)` as these two always get called in pairs. However, this requires a some disentangling of the database and the access to it in beforehand.

I considered this, but in the end I personally felt that responsibility for maintaining the consistency of the database belongs to the Database itself.  Personally, I think eliminating the mixed strings / interval for the Database, and only having it export an interface that consists of Intervals will make the source clearer in the future.

But, it's a judgement call and I see no issue with changing the semantics of validate to include the concept that it can reject intervals without throwing.

**Comment by sruffell** [comment]
> Heres the patch I plan to apply for the fix instead:

```
===================================================================
--- src/commands/CmdStart.cpp	(revision ea33b29986850d9007f2404d7e7030409db85fd4)
+++ src/commands/CmdStart.cpp	(date 1597309652832)
@@ -51,10 +51,11 @@
   }
 
   journal.startTransaction ();
-  validate (cli, rules, database, interval);
-  database.addInterval (interval, verbose);
-  journal.endTransaction ();
-
+  if (validate (cli, rules, database, interval))
+  {
+    database.addInterval (interval, verbose);
+    journal.endTransaction ();
+  }
   if (verbose)
   {
     std::cout << intervalSummarize (database, rules, interval);
```

I think it may be clearer here to add a rollback / reset idea to the Journal here so this function cannot exit with it in a started state.  Then that way if addInterval is called, journal.endTransaction() would be called, but otherwise, journal.resetTransaction() could be called instead.

**Comment by lauft** [comment]
> For example, what should the behavior of the following be?
```
Interval interval {"sopd", "eopd"}
database.addInterval (interval, true);
interval.tag ("foo");
database.addInterval (interval, true);
```
The first `database.addInterval` adds the interval to the database (given that there is no other overlap).
The second `database.addInterval` tries to add an interval with the same borders to the database, which should be rejected in the interval overlap detection in `autoAdjust`.

**Comment by lauft** [comment]
> I think it may be clearer here to add a rollback / reset idea to the Journal here so this function cannot exit with it in a started state. Then that way if addInterval is called, journal.endTransaction() would be called, but otherwise, journal.resetTransaction() could be called instead.

Currently the command ends and the application exists, so all state is destroyed. As this is just "add one interval" or "add no interval", there is nothing to roll back in the latter case.

Yes, a `journal.resetTransaction()` would be cleaner and clearer, but as mentioned above this is just a hacky fix to restore the previous behaviour. I do not want to go down this road too far, especially if the general solution I have in mind will go in a different direction...

**Comment by sruffell** [comment]
> The first `database.addInterval` adds the interval to the database (given that there is no other overlap).
> The second `database.addInterval` tries to add an interval with the same borders to the database, which should be rejected in the interval overlap detection in `autoAdjust`.

Sounds like we're in agreement then that ultimately the database should have validation routines within it.  I thought you were saying that the validation routines should be the users of the database interface, and not the other way around.

**Comment by lauft** [comment]
> Sounds like we're in agreement then that ultimately the database should have validation routines within it.

Well, this depends what you define as 'the database'. To me, there is the low level database, which simply stores strings in textfiles. Wrapped around this is a first layer which adds the interval abstraction. Here we query intervals from the database and hand them over for storage. Currently that's all, and this is more or less hidden in the functions defined in `timew.h` partially implemented in class `Database` itself.

So the commands use "the database" which should do validation, but not at its lowest level.

**Comment by sruffell** [comment]
Closing. I'll base my other branches after you commit your change.

**Pull request closed without merging**